internal/pkg/controllers/profilerecorder/profilerecorder.go

```diff
@@ -0,0 +1,249 @@
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/security-profiles-operator/api/seccompprofile/v1alpha1"
+	"sigs.k8s.io/security-profiles-operator/internal/pkg/config"
+)
+
+const (
+	// default reconcile timeout.
+	reconcileTimeout = 1 * time.Minute
+
+	errGetClient         = "cannot get client connection"
+	errGetNode           = "cannot get node object"
+	errGetPod            = "cannot get pod"
+	errInvalidAnnotation = "invalid Annotation"
+
+	reasonProfileRecording      event.Reason = "SeccompProfileRecording"
+	reasonProfileCreated        event.Reason = "SeccompProfileCreated"
+	reasonProfileCreationFailed event.Reason = "CannotCreateSeccompProfile"
+)
+
+type RecorderReconciler struct {
+	client      client.Client
+	log         logr.Logger
+	record      event.Recorder
+	nodeAddress string
+	podsToWatch map[string]string
+}
+
+func Setup(ctx context.Context, mgr ctrl.Manager, l logr.Logger) error {
+	const name = "profilerecorder"
+	c, err := client.New(mgr.GetConfig(), client.Options{})
+	if err != nil {
+		return errors.Wrap(err, errGetClient)
+	}
+
+	node := &corev1.Node{}
+	if c.Get(
+		context.Background(),
+		client.ObjectKey{Name: os.Getenv(config.NodeNameEnvKey)},
+		node,
+	) != nil {
+		return errors.Wrap(err, errGetNode)
+	}
+
+	nodeAddress := ""
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			l.Info("Setting up profile recorder", "Node", addr.Address)
+			nodeAddress = addr.Address
+			break
+		}
+	}
+
+	if nodeAddress == "" {
+		return errors.New("Unable to get node's internal Address")
+	}
+
+	reconiler := RecorderReconciler{
+		client:      mgr.GetClient(),
+		log:         l,
+		nodeAddress: nodeAddress,
+		record:      event.NewAPIRecorder(mgr.GetEventRecorderFor(name)),
+		podsToWatch: map[string]string{},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithEventFilter(predicate.And(
+			resource.NewPredicates(reconiler.isPodWithTraceAnnotation),
+			resource.NewPredicates(reconiler.isPodOnLocalNode),
+		)).
+		For(&corev1.Pod{}).
+		Complete(&reconiler)
+}
+
+func (r *RecorderReconciler) isPodOnLocalNode(obj runtime.Object) bool {
+	p, ok := obj.(*corev1.Pod)
+
+	if !ok {
+		return false
+	}
+
+	return p.Status.HostIP == r.nodeAddress
+}
+
+func (r *RecorderReconciler) isPodWithTraceAnnotation(obj runtime.Object) bool {
+	p, ok := obj.(*corev1.Pod)
+
+	if !ok {
+		return false
+	}
+
+	return p.Annotations[config.SeccompProfileRecordAnnotationKey] != ""
+}
+
+func (r *RecorderReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
```

Comment:
How about adding events to the reconcile loop when there are failures with a warning? For example `unable to parse annotation`

Reply:
Yep, added an event for that special use case.